File: sdks/python-mcp/tests/mcp_authoring/scenario.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ContentBlockText(StrictModel):
    type: Literal["text"]
    text: str


class ContentBlockImage(StrictModel):
    type: Literal["image"]
    data: str
    mimeType: str


class ContentBlockResource(StrictModel):
    type: Literal["resource"]
    resource: dict[str, Any]


ContentBlock = ContentBlockText | ContentBlockImage | ContentBlockResource


class NudgeSpec(StrictModel):
    kind: Literal["low-balance", "cycle-ending", "approaching-limit"]
    message: str


class ResponseOptions(StrictModel):
    text: str | None = None
    nudge: NudgeSpec | None = None
    units: float | None = None


class HandlerRespond(StrictModel):
    kind: Literal["respond"]
    data: Any
    options: ResponseOptions | None = None
    emit: list[ContentBlock] | None = None


class HandlerGate(StrictModel):
    kind: Literal["gate"]
    reason: str | None = None


class HandlerThrow(StrictModel):
    kind: Literal["throw"]
    message: str


Handler = HandlerRespond | HandlerGate | HandlerThrow


class Limits(StrictModel):
    withinLimits: bool
    remaining: float | None = None
    plan: str | None = None
    creditBalance: float | None = None
    checkoutUrl: str | None = None
    activationRequired: bool | None = None
    confirmationUrl: str | None = None
    plans: list[dict[str, Any]] | None = None
    balance: Any | None = None
    product: Any | None = None
    meterName: str | None = None


class ToolScenario(StrictModel):
    name: str = Field(min_length=1)
    title: str | None = None
    description: str | None = None
    inputSchema: dict[str, Any] | None = None
    args: dict[str, Any]


class Scenario(StrictModel):
    tool: ToolScenario
    product: str = Field(min_length=1)
    customerRef: str = Field(min_length=1)
    customerRefSource: Literal["hook", "toolArgs"]
    usageType: str | None = None
    limits: Limits
    handler: Handler


class UsageProjection(StrictModel):
    outcome: Literal["success", "paywall", "fail"]
    actionType: str
    units: float
    productRef: str
    customerRef: str
    metadata: dict[str, str]


class ToolResult(StrictModel):
    content: list[dict[str, Any]]
    structuredContent: Any | None = None
    isError: bool | None = None
    meta: dict[str, Any] | None = Field(default=None, alias="_meta")


class Observation(StrictModel):
    toolResult: ToolResult
    usage: list[UsageProjection]


def parse_scenario(args: dict[str, Any]) -> Scenario:
    return Scenario.model_validate(args)


def parse_observation(result: object) -> Observation:
    return Observation.model_validate(result)

File: sdks/rust-mcp/tests/common/scenario.rs
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ToolScenario {
    pub name: String,
    pub title: Option<String>,
    pub description: Option<String>,
    pub input_schema: Option<serde_json::Map<String, Value>>,
    pub args: serde_json::Map<String, Value>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "kind", deny_unknown_fields)]
pub enum HandlerSpec {
    #[serde(rename = "respond")]
    Respond {
        data: Value,
        options: Option<Value>,
        emit: Option<Vec<Value>>,
    },
    #[serde(rename = "gate")]
    Gate { reason: Option<String> },
    #[serde(rename = "throw")]
    Throw { message: String },
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Scenario {
    pub tool: ToolScenario,
    pub product: String,
    pub customer_ref: String,
    pub customer_ref_source: String,
    pub usage_type: Option<String>,
    pub limits: Value,
    pub handler: HandlerSpec,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct UsageProjection {
    pub outcome: String,
    pub action_type: String,
    pub units: f64,
    pub product_ref: String,
    pub customer_ref: String,
    pub metadata: serde_json::Map<String, Value>,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct Observation {
    pub tool_result: Value,
    pub usage: Vec<UsageProjection>,
}

pub fn parse_scenario(args: Value) -> Scenario {
    serde_json::from_value(args).expect("scenario")
}

pub fn parse_observation(result: Value) -> Observation {
    serde_json::from_value(result).expect("observation")
}
